catch up with upstream

doccaz opened the pull request:
Merge the latest commits and catch up with upstream.

CendioOssman committed:
Handle all settings via UI.getSetting()

Makes sure everything behaves the same way, even if there is no visible
UI for a settings.

CendioOssman committed:
Merge UI startup in to a single routine

Makes it easier to see how things are connected.

CendioOssman committed:
Make it easier for downstream to modify settings

Expose a simple and stable API to override default settings, and force
settings that users shouldn't be able to change.

CendioOssman committed:
Handle disabling settings without label

ghost commented:
## DeepCode's analysis on [#7485e8](https://www.deepcode.ai/app/gh/doccaz/noVNC/a1409b7eec1b97471ce4f05f641aa43ce5e9d2a9/doccaz/noVNC/7485e82b72d4d1356d95ecca2d109cbf49908b9d/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=2&) found:

- :warning: **1** warning, :information_source: **2** minor issues. :point_down:
- :heavy_check_mark: **4** issues were fixed.
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">Two subsequent branches of an if statement have duplicate bodies (here and here). This may be caused by a copy-paste error. If this usage is intentional, consider merging the branches to avoid code duplication. Occurrences: <ul><li><a href="https://github.com/doccaz/noVNC/blob/7485e82b72d4d1356d95ecca2d109cbf49908b9d/core/rfb.js#L2287">rfb.js:2287</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/doccaz/noVNC/a1409b7eec1b97471ce4f05f641aa43ce5e9d2a9/doccaz/noVNC/7485e82b72d4d1356d95ecca2d109cbf49908b9d/pr/_/%2Fcore%2Frfb.js/javascript%2Fdc_interfile_project%2FDuplicateIfBody/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">resize expects an object in argument 1 and might throw a TypeError on passing a primitive type, but gets number from number literal 4.  Occurrences: <ul><li><a href="https://github.com/doccaz/noVNC/blob/7485e82b72d4d1356d95ecca2d109cbf49908b9d/tests/test.tight.js#L41">test.tight.js:41</a></li> <li><a href="https://github.com/doccaz/noVNC/blob/7485e82b72d4d1356d95ecca2d109cbf49908b9d/tests/test.tight.js#L126">test.tight.js:126</a></li> <li><a href="https://github.com/doccaz/noVNC/blob/7485e82b72d4d1356d95ecca2d109cbf49908b9d/tests/test.rre.js#L53">test.rre.js:53</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/doccaz/noVNC/a1409b7eec1b97471ce4f05f641aa43ce5e9d2a9/doccaz/noVNC/7485e82b72d4d1356d95ecca2d109cbf49908b9d/pr/_/%2Ftests%2Ftest.tight.js/javascript%2Fdc_interfile_project%2FExpectsObjectDislikesPrimitive%2Ftest/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">Signature mismatch: the implementation of ../vendor/pako/lib/zlib/inflate.js.inflateInit has one parameter but two or more arguments are provided in this call. Remove the unnecessary arguments. Occurrences: <ul><li><a href="https://github.com/doccaz/noVNC/blob/7485e82b72d4d1356d95ecca2d109cbf49908b9d/tests/test.deflator.js#L13">test.deflator.js:13</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/doccaz/noVNC/a1409b7eec1b97471ce4f05f641aa43ce5e9d2a9/doccaz/noVNC/7485e82b72d4d1356d95ecca2d109cbf49908b9d/pr/_/%2Ftests%2Ftest.deflator.js/javascript%2Fdc_interfile_project%2FWrongNumberOfArgs%2Ftest/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/doccaz/noVNC/a1409b7eec1b97471ce4f05f641aa43ce5e9d2a9/doccaz/noVNC/7485e82b72d4d1356d95ecca2d109cbf49908b9d/pr/_/%2F/code/?utm_source=gh_review&c=0&w=1&i=2&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=doccaz)
#### 👉 The DeepCode service and API will be deprecated in August, 2021. [Here](https://www.deepcode.ai/blog/migrate-to-snyk-code) is the information how to migrate. Thank you for using DeepCode 🙏 ❤️ !
If you are using our plugins, you might be interested in their successors: [Snyk's JetBrains plugin](https://snyk.co/udpkq) and [Snyk's VS Code plugin](https://snyk.co/udpkr).

samhed committed:
Don't crash if we can't use localStorage

Our settings are not a fatal requirement, we can fall back on the
default values if they can't be accessed. A scenario where we've seen
this happen is when cookies are disabled in the browser. It seems
localStorage is disabled along with cookies in these settings.

So, lets log a message about the failure and otherwise silently
continue in this case.

Fixes issue #1577.

CendioOssman committed:
Change rQslice() to rQpeekBytes()

We don't need any full slice functionality, so let's change this to
better march rQpeek8() and rQshiftBytes().

CendioOssman committed:
Switch Display.flush() to use a promise

That is the modern way to handle operations that cannot complete
immediately.

CendioOssman committed:
Stop direct access to socket buffer

Use proper accessor functions instead of poking around in internal
buffers.

CendioOssman committed:
Stop exposing Websock queue length

Callers should be using rQwait() to ensure sufficient data is present,
and not poke around in the internal buffering.

CendioOssman committed:
Remove Websock implicit read length

Callers should be properly aware of how much data they need, as they
need to call rQwait() first to ensure the data is present.

CendioOssman committed:
Reduce kept state in JPEG decoder

We don't have to keep track of this much data between rects, so
restructure things to make it more simple. This allows the JPEG parsing
code to be a pure function which only depends on the input.

CendioOssman committed:
Remove Base64 WebSocket remnants

There is no encoding/decoding in modern WebSockets, so let's clean up
some of the old crud that no longer serves a purpose.

CendioOssman committed:
Always return copy of data from socket

We don't know how long the caller will hang on to this data, so we need
to be safe by default and assume it will kept indefinitely. That means
we can't return a reference to the internal buffer, as that will get
overwritten with future messages.

We want to avoid unnecessary copying in performance critical code,
though. So allow code to explicitly ask for a shared buffer, assuming
they know the data needs to be consumed immediately.

CendioOssman committed:
Move WebSocket queue index reset to receive

It's more robust to do this just before we need the space, rather than
assume when the queue will be read and adjust things right after.

CendioOssman committed:
Fix Websock send tests

Avoid poking around in the internals and instead test what is actually
sent out on the WebSocket.

CendioOssman committed:
Add Websock send queue helpers

Callers shouldn't be poking around directly in to the send queue, but
should use accessor functions like for the read queue.

CendioOssman committed:
Remove bad Websock mock in tests

This small object will not properly fake a Websock in more complex
cases, so let's avoid it and create a real Websock instead.

CendioOssman committed:
Return a copy of the data from FakeWebSocket

The caller might hang on to the data for multiple calls, so we make sure
the returned buffer might not get overwritten.

CendioOssman committed:
Add missing tests for message encodings

All of these functions should have units tests, even if they are fairly
minimal.

CendioOssman committed:
Make extended clipboard tests independent

Let's test the full final result instead of assuming specific internal
calls.

CendioOssman committed:
Implicitly flush Websock if needed

Callers shouldn't have to deal with the internal buffering limits of
Websock, so implicitly flush the buffer if more room is needed.

CendioOssman committed:
Check that decoders consume all data

This is extra important in the tests where we expect no changes to the
display, as otherwise we can't tell the difference between success and a
decoder that is simply waiting for more data.

CendioOssman committed:
Return unsigned values from rQshift32()

This is what we almost always want, and this makes it consistent with
rQshift8() and rQshift16().

CendioOssman committed:
Avoid internal variables in recv queue tests

Makes for more robust and realistic tests.

CendioOssman committed:
Properly decode ExtendedDesktopSize fields

We are expected to preserve these and use them in our requests back to
the server. We can't do that if we don't actually decode them correctly.

CendioOssman committed:
Don't send SetDesktopSize too early

We don't know the server layout yet, so we can't preserve the screen id
or flags yet at this point. Move it until after we've parsed everything.

CendioOssman committed:
Make ExtendedDesktopSize tests more realistic

Send real messages and avoid poking around in internals, as we weren't
testing things correctly that way.

CendioOssman committed:
Use proper socket helpers for FBU header

Let's not duplicate this stuff when we have convenience functions.

CendioOssman committed:
Merge branch 'testsfix' of github.com:CendioOssman/noVNC

CendioOssman committed:
Don't split large WebSocket data in tests

It takes too much time and can make the tests fail.

CendioOssman committed:
Merge branch 'websock' of https://github.com/CendioOssman/noVNC

samhed committed:
Merge pull request #1710 from novnc/localStorage

Don't crash if we can't use localStorage

otthou committed:
Add QEMU Led Pseudo encoding support

Previously, num-lock and caps-lock syncing was performed on a best effort basis by qemu.
Now, the syncing is performed by no-vnc instead. This allows the led state syncing to work
in cases where it previously couldn't, since no-vnc has with this extension knowledge of both
the remote and local capslock and numlock status, which QEMU doesn't have.

CendioOssman committed:
Use proper argument to deflateInit()

This was an accidental copy error from inflator.js. The second argument
to deflateInit() is the compression level, not the window bits.

We have not strong opinions on an appropriate level, so stick to the
default.

CendioOssman committed:
Remove unused argument to inflateInit()

There is just one argument to inflateInit(). It is inflateInit2() that
takes two arguments.

Since this argument was never used, let's just remove it and keep the
existing behaviour.

CendioOssman committed:
Fix zlib level change in clipboard tests

The compression level got changed in 01bb36d4, but the tests weren't
updated to follow this change.

CendioOssman committed:
Correctly handle "none" auth on old servers

There is no security result for the "none" authentication until RFB 3.8.
This got broken by mistake in 5671072.

CendioOssman committed:
Correctly handle legacy security rejections

The code comment of this code was entirely incorrect, but the commit
message for 5671072 when it was added was correct. I.e. there is a
result, but not a reason.

Adjust the unit tests to make sure this doesn't regress again.

CendioOssman committed:
Fix bad indentation

CendioOssman committed:
Merge branch 'qemu_ledstate_pseudo_encoding' of https://github.com/otthou/noVNC

afbjorklund committed:
Add the possibility to listen on a specific host

For instance, for listening only on "localhost"

That is, bind on 127.0.0.1 instead of 0.0.0.0

afbjorklund committed:
Document default port applies to all interfaces

afbjorklund committed:
Remove comment about websockify command arguments

CendioOssman committed:
Merge branch 'listen-host' of https://github.com/afbjorklund/noVNC

CendioOssman committed:
Handle immediate connection errors

The browser might throw an exception right away if there is something it
doesn't like with our connect attempt. E.g. using a non-TLS WebSocket
from a TLS web page.

CendioOssman committed:
Handle relative paths in novnc_proxy

websockify changes the working directory before it starts looking for
files, so we must give it relative paths for things to work reliably.

CendioOssman committed:
Remove unused npm dependencies

These should have been removed as part of 890cff9.

CendioOssman committed:
Upgrade base snap to Ubuntu 22.04

Ubuntu 18.04 base snap is no longer supported, so switch to the
currently newest one.

CendioOssman committed:
Don't include ES6 module versions in npm package

The npm package is supposed to be for CommonJS usage, so only package
that to avoid confusion. This has become an issue now that nodejs
supports ES6 modules, where users are accidentally trying to import the
wrong files and get errors.

CendioOssman committed:
Increase test timeout for Chrome on Windows

There is some bug in Chrome 119+ on some systems, where it takes
forever for the first readback from a canvas, timing out the first
test that does this.

Work around the issue by increasing the timeout on that platform until
Chrome manages to resolve the issue.

CendioOssman committed:
Handle broken Oculus browser keyboard events

It sets KeyboardEvent.key to "Unidentified" for all non-character keys,
which means we must ignore it and use the legacy handling to figure out
the key pressed.

CendioOssman committed:
Fix key event debug output

Fix for a0b7c0dac5359e4002e7f1d946e60e2eb9b4a54e.

sbungartz committed:
Avoid exception when cursor was removed from DOM already

samhed committed:
Make non-HTTPS message more harsh

As browsers are placing more and more new functionality as
secure-context only, we need to prepare users for more problems. I find
it likely that we will disable non-HTTPS connections in the future.

CendioOssman committed:
Update github actions to latest versions

Primarily to avoid the versions that are now deprecated, but also update
actions/upload-artifact to keep us up to date.

CendioOssman committed:
Fix import of "commander"

The default import was deprecated ages ago, and in v12 it has now
finally been changed in a breaking way.

Change the code to import things the proper way.

CendioOssman committed:
Merge branch 'actions' of github.com:CendioOssman/noVNC

samhed committed:
Merge pull request #1834 from sbungartz/avoid-exception-when-disconnecting-after-dom-morph

Avoid exception when cursor was removed from DOM already

Commit:
Clipboard: handle multiple CR+LF

samhed committed:
Migrate deprecated eslint config to to new format

The .eslintrc and .eslintignore formats are deprecated. The new format
uses a single eslint.config.js (or .mjs) file at the top.

Commit:
Update zh_CN.po (#1851)

Update Chinese translation

kosmasgiannis committed:
Updated el.po

samhed committed:
Merge pull request #1853 from kosmasgiannis/gr20240424

Updated greek translations

samhed committed:
Cleanup "no-console" eslint rules

Removes unexpected exceptions and clarifies where we want to avoid
console calls.

mentha committed:
Implement gradient filter of tight decoder, fixing issue #1767

This commit is a basic implementation of the gradient filter required by
qemu `lossy` option.

CendioOssman committed:
Add unit tests for Tight gradient filter

CendioOssman committed:
Remove Twitter links

These are not updated anymore as they are not under the control of the
current team.

CendioOssman committed:
Update translation template file

CendioOssman committed:
noVNC 1.5.0 beta

CendioOssman committed:
Upgrade to websockify 0.12.0 in snap package

CendioOssman committed:
Remove Chrome timeout workaround

This is a revert of fca48df85d394b2c1e9d26ed6e9d10d33db9f4f4. The issue
seems to be fixed in the current version of Chrome, so let's keep things
simple again.

CendioOssman committed:
Don't include missing translation in .js

It just adds size and confusion. Instead, omit any lines where no
translation is available.

CendioOssman committed:
Update Swedish translation

CendioOssman committed:
Update json files for new translations

CendioOssman committed:
noVNC 1.5.0

any1 committed:
Add H.264 decoder

This adds an H.264 decoder based on WebCodecs.

samhed committed:
Use theme-color to color address bar in browsers

This makes the address bar on mobile browsers match the background. Note
that it requires a valid certificate and a non-dark mode set on the
device. Not supported on desktop browsers.

leedagee committed:
Interrupt AltGr sequence detection on focus lost, fixes #1880

CendioOssman committed:
Manually load sinon and chai

karma-sinon-chai is not compatible with Chai 5+, and Karma is no longer
being updated.

Load sinon and chai manually instead, until we can have a long term plan
in place.

CendioOssman committed:
Stop creating sinon sandbox early

sinon might not be loaded at this point, which can cause tests to fail.

We could create the sandbox in one of the hooks instead, but let's
remove the sandbox completely to stay consistent with our other tests.

CendioOssman committed:
Merge branch 'altgr-seq-interrupt' of https://github.com/leedagee/noVNC

CendioOssman committed:
Add test for AltGr abort on blur

CendioOssman committed:
Let browser construct URL string for us

Likely a lot safer for corner cases than us trying to figure this out
ourselves.

CendioOssman committed:
Allow relative WebSocket URLs

This can be very useful if you have multiple instances of noVNC, and you
want to redirect them to different VNC servers.

The new default settings will have the same behaviour as before for
systems where noVNC is deployed in the root web folder.

CendioOssman committed:
Load settings from web server

Make it even easier to customize things by loading the settings from
separate configuration files.

any1 committed:
Add unit tests for H.264 decoder

markpeek committed:
Add Zlib encoding

CendioOssman committed:
Fix host and port via query string

We need to call initSetting() even if we don't have any interesting
default to set, as that is what checks if values have been provided as a
query string.

Fixes 96c76f7.

Bobini1 committed:
Fix sQpushBytes sending the beginning of the array multiple times

CendioOssman committed:
Let fake WebSocket handle large sends

Dynamically grow the recorded send buffer if the test needs to send a
lot of data.

CendioOssman committed:
Fix tests for large WebSocket sends

These failed to test that the data was correctly split as they only
checked the first chunk transmitted.

Use random values to avoid the risk of aligning our test data with the
split boundaries and hence allowing false positives.

CendioOssman committed:
Merge branch 'webcodec-h264' of https://github.com/any1/noVNC

CendioOssman committed:
Merge branch 'ui_init' of github.com:CendioOssman/noVNC

dim5x committed:
Fix typos in Russian translation

CendioOssman committed:
Fix handling of VideoDecoder.isConfigSupported()

It returns an object with details, not just a simple boolean.

CendioOssman committed:
Use common H.264 check in tests

Avoid duplicating this logic in multiple places.

CendioOssman committed:
Do a real H.264 test decode to determine support

Firefox is buggy and reports support for H.264 but then throws errors
once we actually try to decode things. Detect this early by doing a
quick test decode of a single frame.

CendioOssman committed:
Handle exceptions from VideoDecoder.flush()

These are not supposed to happen according to the specification, but
Firefox has some bug and throws them anyway.

CendioOssman committed:
Detect broken Firefox H.264 decoder

The Firefox H.264 decoder on Windows might simply just refuse to deliver
any finished frames. It also doesn't deliver any errors.

Detect this early by expecting a frame after flush() has completed.

CendioOssman committed:
Raise JavaScript version requirement

So that we can use await at module top level.

wxtewx committed:
Update zh_CN.po

CendioOssman committed:
Consistently use "sentence case" style

Try to be more consistent in how we capitalize things. Both the "Title
Case" and "Sentence case" styles are popular, so either would work.
Google and Mozilla both prefer "Sentence case", so let's follow them.

CendioOssman committed:
Also adjust to "sentence case" in translations

This would have resolved itself automatically on the next translation
update, but this commit will reduce unnecessary noise in that change.

CendioOssman committed:
Merge branch 'master' of https://github.com/wxtewx/noVNC

CendioHalim committed:
Use MouseEvent.buttons for button state tracking

Instead of keeping track of button states ourselves by looking at
MouseEvent.button, we can use the MouseEvent.buttons which already
contains the state of all buttons.

nnn1590 committed:
Update Japanese translation

CendioOssman committed:
Replace node-getopt with commander for args

node-getopt isn't maintained and nodejs has started complaining about
deprecated features in it.

CendioOssman committed:
Replace po2json with pofile

The former doesn't seem to be properly maintained and nodejs gives
deprecation warnings.

samhed committed:
Fix resizes back to initial remote session size

Since the expected client size wasn't updated when the browser window
resized, noVNC didn't resize the canvas properly when going back to
the exact same dimensions.

Fixes issue #1903

samhed committed:
Comment different resize functions in rfb.js

CendioHalim committed:
Add support for forward and back mouse buttons

This commit implements the extendedMouseButtons pseudo-encoding, which
makes it possible to use the forward and back mouse buttons.

CendioHalim committed:
Move mouse event help functions to broader scope

These functions can be used elsewhere in the tests. We want to use these
in the dragging tests in the future instead of directly calling private
methods.

CendioHalim committed:
Split button click with dragging test

CendioHalim committed:
Dispatch mouse events in dragging unit tests

This makes our tests reflect the real world better, as we now send real
mouse events instead of calling private methods directly.

CendioHalim committed:
Check for correct button events in dragging tests

Previously, these unit tests did not check which events were sent to the
server, only how many events were sent. This commit adds checks to see
that the expected button events are sent.

CendioHalim committed:
Move gesture event help functions to broader scope

This is needed if we want to test gestures with dragging.

CendioHalim committed:
Add unit test for wheel + buttons pressed

samhed committed:
Standardize on 4 space-indentation in CSS files

This is what we use in every other file in noVNC. It also much more
common for a CSS file in general. By standardizing on 4 spaces we can
avoid indentation mistakes.

samhed committed:
Increase padding of buttons and inputs

Gives them a more modern and spacious look.

samhed committed:
Add margin between label and input in noVNC_panel

To make stuff feel less cramped, lets add some margin here.

As of comitting this, it only affects the logging-level select dropdown
in the settings, but this is a general rule of thumb. It doesn't apply
to checkboxes or radios since they have a margin by default, and their
label to the left.

samhed committed:
Create CSS variables for common noVNC-colors

samhed committed:
Standardize on 6px or 12px border-radius

This results in a few things becoming slighly more rounded, for example
the controlbar, the settings panel and buttons/inputs. Increased
rounding gives a more friendly feel.

samhed committed:
Get rid of gradients on buttons and inputs

Lets make things more flat and modern.

samhed committed:
Differentiate buttons from text inputs

By making buttons grey with bold text, they are easy to distinguish from
text inputs.

samhed committed:
Redesign select dropdown arrow

Makes it more of a "V"-shape rather than a triangle, suits better in the
new spacier select-buttons.

samhed committed:
Redesign checkboxes and radiobuttons

Makes them bigger and gets rid of their borders. The change also allowed
for some shared styling between them.

samhed committed:
Add indeterminate styling to checkboxes and radios

This is used when the control is neither checked or unchecked.

samhed committed:
Rework how buttons react to :hover and :active

Instead of having two different types of effects (hover had a different
color, and active had a 3d-effect simulating a pressed button), we now
have an increasing activation-level. That means the button goes a bit
dark for hover, and then even darker when pressed.

There is also a variant that goes lighter for each activation level,
that can be used when the initial color is dark.

With this change, we can get rid of special :hover and :active styling
for the connect button and the control bar buttons. We can use the same
activation level principle for all buttons.

samhed committed:
Make buttons flat by removing borders

Gives a more clean look that fits well with the new checkboxes and
radios. The old border was mostly used to contribute to a 3d-effect,
that was used for :active. That :active-styling has been replaced by
activation levels.

samhed committed:
Fix :disabled styling of file-selector-button

By applying the rule to the button within the input, we effectively
applied the opacity twice - making the button almost disappear. Applying
the opacity to the input element is enough.

samhed committed:
Put specific :disabled rules with its element

It makes more sense to group rules per element type.

samhed committed:
Move general :disabled rules to common top section

Start with general stuff, followed by specific things.

samhed committed:
Support 'disabled' attribute on labels

Note that the :disabled selector only works on inputs, buttons and the
like.

The current method of applying .noVNC_disabled to the settings
labels is still used. This support is added mostly for completeness.

Note that when a label wraps an input, only the label should have the
disabled attribute. Otherwise the effect applies twice to the input.

samhed committed:
Capitalize CSS section headings

Makes them easier to distinguish from regular comments.

samhed committed:
Keep CSS for different input elements together

By moving the CSS for select buttons to the bottom, we keep a more
logical ordering of the elements.

samhed committed:
Avoid 2 borders when focus-visible on text inputs

By having the focus-visible outline overlapping the regular border
things look a bit more sane on text input elements.

samhed committed:
Add some sane default CSS for textareas

samhed committed:
Remove number picker's increase/decrease buttons

We can't style them, and they don't fit the noVNC CSS style - we are
better off without them.

samhed committed:
Standardize on ellipsis-type text-overflow

samhed committed:
Add minimum width to buttons

This ensures they aren't too small, even if the text label is short.

samhed committed:
Pointer cursor on buttons & grab on sliders

This makes buttons and slider stand out more.

samhed committed:
Add styling for color pickers

Note that no color picker elements are currently in use, this is for
completeness.

samhed committed:
Make interface airier by increasing line-height

Modern interfaces are less cramped, this makes noVNC feel more up to
date.

Note that this required some adjustments on noVNC_headings and
noVNC_connect_button since the text now takes up more height than the
images.

samhed committed:
Add styling for toggle switches

These are a type of checkbox that is suitable for ON/OFF-type switches.

samhed committed:
Utilize toggle switch in settings

These settings are well suited to use toggle switches. This makes these
settings more approachable and user-friendly.

samhed committed:
Add styling for checked options in select boxes

samhed committed:
Add styling for multi-select lists

This is a type of select box that doesn't appear like a button, but more
like a textarea that lists options. It is not currently used, but added
for completeness.

samhed committed:
Make range slider thumbs circular

Fits better with the new slightly rounded and spacious style. The track
was made slightly thicker to ensure proper centering of the new thumb.

samhed committed:
Only color the left part of the range track

Makes it easier to envision the value is selected.

samhed committed:
Make text in panel headings bold

Makes the heading pop a bit more in the new airier layout.

CendioHalim committed:
Don't send mouse events when dragging viewport

We don't want to send any mouse events to the server when dragging the
viewport. Instead, we treat them as a client-only operation.

samhed committed:
Fix spacing between elements in dialogs

Dialogs have had text inputs and buttons cramped together without space
between, this fixes that.

CendioHalim committed:
Add tests for dragging with gestures

There were no test for viewport dragging using gesture previously, so
let's add some.

Note that there currently are some viewport dragging behaviours that we
don't want to have, so some tests have commented out what our desired
behaviour should be.

samhed committed:
Tweak design of noVNC connect button

Make the color contrast with the background and the button more rounded.
The goal is to make the button stand out.

CendioHalim committed:
Move mouse move flushing to separate function

CendioHalim committed:
Flush mouseMove when initiating viewport dragging

We want to flush pending mouse moves before we initiate viewport
dragging.

Before this commit, there were scenarios where the _mouseButtonMask
would track a released button as being down.

CendioOssman committed:
Merge branch 'mouse-buttonstate-remove' of https://github.com/CendioHalim/noVNC

CendioHalim committed:
Move sendFbuMsg() to broader scope

This is needed if we want to use this function elsewhere in our tests.

CendioOssman committed:
Merge branch 'extra-mouse-buttons' of https://github.com/CendioHalim/noVNC

CendioOssman committed:
Also test Ctrl+AltGr, as that is what browsers use

Modern browsers now send the odd sequence of Ctrl+AltGr, rather than the
raw Ctrl+Alt, or the fully adjusted just AltGr.

Make sure we have a test for this scenario and don't break it.

samhed committed:
Fix settings panel layout on small screens

Both labels and inputs protruded outside the panel on for example a
phone in portrait mode. This commit fixes that by allowing wrapping and
setting a max-width.

Since the --input-xpadding variable is now used in two different CSS
files, it was moved to constants.css.

CendioZeijlon committed:
Remove bold styling tags in vnc.html

We aren't emphasizing important information with bold tags anywhere
else, so we shouldn't do it here either.

samhed committed:
Add exceptions for CSS validator false positives

Some new CSS incorrectly give errors from validator.w3.org. Issues were
opened in that repo, so hopefully we can remove these exceptions soon.

I searched for alternative validators, but couldn't find a different one
that had a simple API like this one.

In order to reliably detect & handle these exceptions we unfortunately
need to make the validator output parsing quite a bit more complicated.

samhed committed:
Make the background of expanded settings lighter

A very slight change to the background color, to make the contrast
better with the light-grey input elements.

CendioZeijlon committed:
Merge branch 'ui_refresh'

CendioZeijlon committed:
Update README.md with ExtendedMouseButtons feature

samhed committed:
Use less technical phrasing in README

CendioOssman committed:
Add helper for ExtendedDesktopSize in tests

CendioOssman committed:
Make resizeSession setting test more realistic

We shouldn't expect a resize request to be sent if the container didn't
change size first.

CendioOssman committed:
Include SetDesktopSize responses in tests

There might be subtle changes in behaviour, so we should mimic what a
real server does.

CendioOssman committed:
Better resize rate limiting

Be more aggressive with resizing, limiting it to once ever 100 ms
instead of after a 500 ms idle period. This gives a more responsive user
experience.

CendioOssman committed:
Merge branch 'resize' of https://github.com/CendioOssman/noVNC

tianzedavid committed:
chore: fix some typos

Signed-off-by: tianzedavid <cuitianze@aliyun.com>

CendioZeijlon committed:
Merge pull request #1934 from tianzedavid/master

chore: fix some typos

CendioZeijlon committed:
Update translation template file

CendioZeijlon committed:
Update Swedish translations

CendioZeijlon committed:
noVNC 1.6.0 beta

CendioZeijlon committed:
Update README.md with H.264 encoding support

voilier-martineke committed:
Update French translation

CendioZeijlon committed:
Upgrade to websockify 0.13.0 in snap package

haroldh committed:
Update Dutch translation

CendioZeijlon committed:
Let browser handle parsing of URLs before relaying

We don't want to assign a path directly to url.pathname that contains a
search query, since this causes '?' at the beginning of the query to be
URL-encoded to '%3F'. Instead use URL() to parse the path for us.

CendioZeijlon committed:
Merge branch 'fix-sub-url-query'

CendioZeijlon committed:
Update generated json files for new translations

CendioZeijlon committed:
noVNC 1.6.0

CendioOssman committed:
Only include valid translations in .json files

Fuzzy translations might be incorrect, and obsolete translations aren't
used anywhere.

samhed committed:
Fix appearance of extra key buttons

Since the extra keys panel is quite narrow in width, a max-width style
resulted in the buttons almost disappearing. That rule was only intended
for elements inside the settings panel.

Broken by commit 14f9ea5880f32f2a4867006d46c8e871942c698e.

Another minor error that is also fixed by this commit is that the
clipboard textarea no longer incorrectly gets a left margin of 6px.

Fixes #1946.

samhed committed:
Update comment reference missed in previous commit

Should have been part of f0a39cd357a5995673149b95951d4c1261b69571

CendioOssman committed:
Close VideoFrame after H.264 detection

Otherwise browser will complain when it is garbage collected.

liaopeiyuan committed:
Fix typo in error-handler.js

Leostruka committed:
Fix typo in Portuguese translation for "Show extra keys"

CendioZeijlon committed:
Merge branch 'master' of github.com:Leostruka/noVNC

CendioOssman committed:
Document new behaviour of host/port/encrypt/path

This changed in 96c76f7, but we forgot to adjust the documentation for
the parameters.

CendioOssman committed:
Document new JSON settings files

This mechanism was added in 438e5b3, but we forgot to document it.

elikoga committed:
Remove deprecated "directories" entry

elikoga committed:
Convert NPM bundle to ES-Module format

CendioZeijlon committed:
Disable setting showDotCursor in RFB constructor

This has been deprecated for around six years now. Let's remove the
deprecation warning and disable setting showDotCursor via the options
parameter.

CendioZeijlon committed:
Merge branch 'obsolete-showDotCursor-rfb-option' of github.com:ThinLinc-Zeijlon/noVNC

Commit:
Add permissions-exclusive async clipboard

Clipboard permissions must be supported, with states "granted" or
"prompt" for both write and read.

Commit:
Add async clipboard module to internal API docs

milotype committed:
Add Croatian translation

CendioOssman committed:
Scan all buffered data looking for JPEG end

This is much more efficient than looking at two bytes at a time.

CendioOssman committed:
Expose length of buffered WebSocket data

Some encodings don't know how much data they need, rather they must
probe the data stream until they find an end marker. Expose how much
data is buffered in order to make this search efficient.

CendioZeijlon committed:
Use Croatian translations

CendioZeijlon committed:
Fix broken Chai import

Chai v6.0.0 introduced a breaking change where file imports now need to
point at 'chai/index.js'. See the corresponding release note.

CendioOssman committed:
Drop Image data once rendered

Helps the browser to free up the memory right away, rather than waiting
until some later cleanup process. At least Firefox can start consuming
gigabytes of memory without this.

tsukasa-au committed:
Add tests for the wakelock feature.

Add tests to for both the `rfb` side (calling into the new wakelock
code), and the new wakelock class (which tracks the desired state and
how to get there).

tsukasa-au committed:
Add wakelock support

Add a new configuration option `keep_device_awake` to allow noVNC to
stop the local display from going to sleep. This is especially useful
with view-only sessions.

This new option has been added to the configuration UI, making it easier
for users to configure. When this option is changed at runtime, we will
request/release the wake lock.

We only hold the view lock while connected to a server. We will also
attempt to reacquire the wakelock if we lost it due to a visibility
change (the tab becoming inactive, or during the transition into/from
fullscreen).

All existing unittests have been run, and the change has been manually
tested in Firefox 145. Additional tests will be added later.

tsukasa-au committed:
Add error state for wakelock testing.

Add an error state to the WakeLockManager state machine. This adds a
transition that can be detected from tests (it otherwise serves no
purpose, and the system immediatly transitions back into the released
state).

tsukasa-au committed:
Expose WakeLockManager state transitions for tests

Dispatch an event on each state transition inside the WakeLockManager.
This gives the unit tests something to synchronise on, allowing us to
write fast, flake-free tests.

Commit:
Disable clipboard button with async clipboard

With async clipboard available, the fallback clipboard textarea adds
mostly confusion. If async clipboard is out right denied, users most
likely don't want to see any clipboard activity.

CendioOssman committed:
Merge branch 'switch-to-es-module-publish-and-use-exports' of github.com:elikoga/noVNC

Commit:
Add Hungarian translation

CendioZeijlon committed:
Remove obsolete checks in novnc_proxy

Our snap packages no longer contain python2-websockify. It's now called
websockify and is found in snap by the WEBSOCKIFY_FROMSYSTEM check.

CendioZeijlon committed:
Use bash's type command to find websockify

Our snap package requires bash to run novnc_proxy, but it doesn't
explicitly set a requirement for the which command to be installed.
Let's therefore use a bash built-in when looking for the websockify
binary.

CendioZeijlon committed:
Merge branch 'clipboard-async' of https://github.com/tobfah/noVNC

Commit:
Display warning prompt before closing tab

The warning prompt is only displayed if there is an active connected
session, when viewOnly is disabled.

CendioZeijlon committed:
Merge branch 'master' of https://github.com/Deni42/noVNC

CendioZeijlon committed:
Update translation template file

CendioZeijlon committed:
Update Swedish translation

CendioZeijlon committed:
Allow publishing to npmjs.com with OIDC

CendioZeijlon committed:
Stop using access tokens when publishing to npmjs

We are using OIDC now instead of the old style access tokens.

CendioZeijlon committed:
Publish with latest npm version

Apparently publishing with OIDC as an auth method only works with npm
11.5.1 and onward.

CendioZeijlon committed:
Use Node version 24 when publishing to npmjs

Node 24 comes with npm > 11.5.1 which we need in order to publish with
OIDC auth.

This also reverts the change in commit e03a54e, which broke the
publishing step.

CendioZeijlon committed:
Remove showDotCursor from docs/API.md

This should have been a part of commit 243d7fd

CendioZeijlon committed:
Remove show_dot from docs/EMBEDDING.md

This should have been a part of commit 243d7fd

nykula committed:
Translate to Ukrainian

MantriHimaVamshi-AidenAI committed:
Fix comment typo in display.js

I noticed a simple typo - "inoder rendering"

samhed committed:
Merge pull request #1984 from tsukasa-au/add-wakelock-support

Add wakelock support

samhed committed:
Merge pull request #2029 from nykula/nykula-patch-1

Translate to Ukrainian

samhed committed:
Merge pull request #2035 from MantriHimaVamshi-AidenAI/patch-1

Fix comment typo in display.js

CendioOssman committed:
Remove deprecated host and port from example

We shouldn't be advertising these if we'd like people to stop using
them.

CendioOssman committed:
Adapt to latest version of eslint

CendioOssman committed:
Remove some unused assignments

Mainly to keep eslint happy with its new recommended rules. But it also
makes the code more clear to not have unused stuff in it.

The des code got a bit more refactoring since eslint was upset about the
final "i++" as it was technically an unused assignment.

Commit:
Enable attaching control bar to top and bottom

Commit:
Make scroll bar and key modifiers flex box

In preparation for horizontal control bar modes

Commit:
Use svg for control bar icon

Simplifies alignment. In preparation for horizontal control bar modes.

Commit:
Remove redundant mobile buttons grouping

There's only one button in the group, and the wrapping div messed with
the flex box layout

samhed committed:
Remove validation exception for "scrollbar-gutter"

This issue has now been fixed in https://validator.w3.org/nu, see:

https://github.com/w3c/css-validator/issues/473

samhed committed:
Fix broken clipboard unittests

We can't compare with a new CustomEvent, lets check the call arguments
instead. The field is also called "text".

dmotte committed:
Only focus username field if empty

https://github.com/novnc/noVNC/pull/2057

zk-wz committed:
Workaround endless loading on Android Chrome

Replace top-level await with then() in browser.js, due to a bug in
chromium browsers.

https://github.com/novnc/noVNC/pull/2053

samhed committed:
Update github actions to latest versions

samhed committed:
Merge pull request #2058 from novnc/github-actions

Update github actions to latest versions

samhed committed:
Update version in package.json to 1.7.0

This was missed when doing the 1.7.0 release.

samhed committed:
Use macOS 26 runner instead of macOS 10.15

This fixes failing tests on Safari.

samhed committed:
Merge pull request #2059 from novnc/macos26

Use macOS 26 runner instead of macOS 10.15

samhed committed:
Merge pull request #2043 from tobfah/enable-horizontal-control-bar

Enable attaching the control bar to top and bottom

ArduFish committed:
Update vnc.html

samhed committed:
Merge pull request #2046 from ArduFish/master

Fix typo
